plugins/gpio/fu-plugin-gpio.c

```diff
@@ -0,0 +1,138 @@
+	return fu_io_channel_write_raw(io_channel,
+				       (const guint8 *)gpio_number_str,
+				       strlen(gpio_number_str),
+				       1500,
+				       FU_IO_CHANNEL_FLAG_SINGLE_SHOT,
+				       error);
+}
+
+static gboolean
+fu_plugin_gpio_set_direction(FuPlugin *self, guint gpio_num, const gchar *direction, GError **error)
+{
+	g_autofree gchar *fn = NULL;
+	g_autofree gchar *devname = g_strdup_printf("gpio%u", gpio_num);
+	g_autoptr(FuIOChannel) io_channel = NULL;
+
+	g_return_val_if_fail(direction != NULL, FALSE);
+
+	fn = g_build_filename("/sys/class/gpio", devname, "direction", NULL);
+	io_channel = fu_io_channel_new_file(fn, error);
+	if (io_channel == NULL)
+		return FALSE;
+	return fu_io_channel_write_raw(io_channel,
+				       (const guint8 *)direction,
+				       strlen(direction),
+				       1500,
+				       FU_IO_CHANNEL_FLAG_SINGLE_SHOT,
+				       error);
+}
+
+static gboolean
+fu_plugin_gpio_process_quirk_entry(FuPlugin *self, const gchar *str, GError **error)
+{
+	guint64 gpio_num = 0;
+	g_auto(GStrv) split = g_strsplit(str, ":", -1);
+
+	if (g_strv_length(split) != 2) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "invalid format, NUMBER:LEVEL, got '%s'",
+			    str);
+		return FALSE;
+	}
+	if (!fu_common_strtoull_full(split[0], &gpio_num, 0, G_MAXUINT, error))
+		return FALSE;
+	if (!fu_plugin_gpio_create_node(self, gpio_num, error))
+		return FALSE;
+	return fu_plugin_gpio_set_direction(self, gpio_num, split[1], error);
+}
+
+static gboolean
+fu_plugin_gpio_process_quirk(FuPlugin *self, const gchar *str, GError **error)
+{
+	g_auto(GStrv) split = g_strsplit(str, ",", -1);
+	for (guint i = 0; split[i] != NULL; i++) {
+		if (!fu_plugin_gpio_process_quirk_entry(self, split[i], error))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_gpio_process_quirks(FuPlugin *self, FuDevice *device, const gchar *key, GError **error)
+{
+	GPtrArray *guids = fu_device_get_guids(device);
+	for (guint i = 0; i < guids->len; i++) {
+		const gchar *guid = g_ptr_array_index(guids, i);
+		const gchar *str;
+		str = fu_context_lookup_quirk_by_id(fu_plugin_get_context(self), guid, key);
+		if (str == NULL)
+			continue;
+		if (!fu_plugin_gpio_process_quirk(self, str, error))
+			return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_gpio_backend_device_prepare_cb(FuPlugin *self,
+					 FuDevice *device,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	return fu_plugin_gpio_process_quirks(self, device, "GpioPrepare", error);
+}
+
+static gboolean
+fu_plugin_gpio_backend_device_cleanup_cb(FuPlugin *self,
+					 FuDevice *device,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	return fu_plugin_gpio_process_quirks(self, device, "GpioCleanup", error);
+}
+
+void
+fu_plugin_init_vfuncs(FuPluginVfuncs *vfuncs)
+{
+	vfuncs->build_hash = FU_BUILD_HASH;
+	vfuncs->init = fu_plugin_gpio_init;
+	vfuncs->prepare = fu_plugin_gpio_backend_device_prepare_cb;
```

Comment:
presumably some plugins may need this for probing too?  Need to think about that at all?

Reply:
Hmm, good question. I suppose it's almost free to add, but my primary usecase was something like turning on a LED, clearing a WR-EN signal or changing the MUX on a host<->BMC.

Reply:
Lets leave this just for prepare for the moment, and we can add another key as-and-when required. It'll be all the same plumbing.